### PR TITLE
Binary optimization: Support Feature Flags #429

### PR DESCRIPTION
**Pull request by saraswatpuneet**
# Goal
The goal of this PR is  to enable specific flags for different runtimes

part of #429 

# Checklist
- [x] Update Frequency node
- [x] Test cargo check for all 4 features ```--features frequency``` , ```--features frequency-rococo-local``` , ```--features frequency-rococo-test``` and ```--features all-frequency-features```
- [x] Test cargo build for all features 
- [x] Run instant sealing mode with feature flag turned on and test frequency rococo-local chain
- [x] Run unit tests
- [x] Test benchmarks 
- [x] Update makefile and scripts 
- [x] Update spec/binary jobs for rococo and mainnet
- [x] Update main and PR jobs to run all feature

### CODE REVIEW

File: .github/workflows/frequency_specs.yml [hunk 1]
```diff
@@ -39,36 +39,38 @@ jobs:
         uses: Swatinem/rust-cache@v1
         with:
```
Comment (saraswatpuneet) [inline .github/workflows/frequency_specs.yml:40]: Updates Here
- Remove need for generating rococo-local binary 
- Create rococo and mainnet binaries separately

File: Cargo.toml [hunk 2]
```diff
@@ -12,3 +9,11 @@ members = [
...
   "common/helpers",
 ]
+
+[profile.release]
+panic = "unwind"
+
+[profile.production]
+inherits = "release"
+lto = true
+codegen-units = 1
```
Comment (saraswatpuneet) [inline Cargo.toml:19]: production profile for releases

File: Makefile [hunk 3]
```diff
@@ -72,3 +72,35 @@ docs:
...
+	cargo build --locked --release --features  frequency-rococo-testnet
+
+build-mainnet:
+	cargo build --locked --release --features  frequency
+
+build-rococo-release:
+	cargo build --locked --features  frequency-rococo-testnet --profile production
+
+build-mainnet-release:
+	cargo build --locked --features  frequency --profile production
```
Comment (saraswatpuneet) [inline Makefile:106]: add some additional make commands, however cargo check/build will still generate full binary, so nothing will break for devs in their current way of working

File: node/Cargo.toml [hunk 4]
```diff
@@ -16,19 +16,25 @@ path = "src/main.rs"
...
 runtime-benchmarks = [
   "frequency-runtime/runtime-benchmarks",
   "frequency-rococo-runtime/runtime-benchmarks",
   "frequency-cli/runtime-benchmarks",
 ]
+frequency = ["frequency-service/frequency", "frequency-cli/frequency"]
+frequency-rococo-local = ["frequency-service/frequency-rococo-local", "frequency-cli/frequency-rococo-local"]
+frequency-rococo-testnet = ["frequency-service/frequency-rococo-testnet", "frequency-cli/frequency-rococo-testnet"]
+all-frequency-features =["frequency-service/all-frequency-features", "frequency-cli/all-frequency-features"]
 try-runtime = ["frequency-runtime/try-runtime", "frequency-rococo-runtime/try-runtime"]
```
Comment (saraswatpuneet) [inline node/Cargo.toml:40]: define 3 features
1. frequency  = mainnet
2. frequency-rococo-testsnet = rococo testnet features 
3. frequency-rococo-local = same as frequency-rococo-test + instant sealing is only included in this feature

File: node/service/src/chain_spec.rs [hunk 5]
```diff
@@ -1,13 +1,22 @@
...
+pub type DummyChainSpec = sc_service::GenericChainSpec<(), Extensions>;
+
+#[cfg(feature = "frequency")]
 pub mod frequency;
+
+#[cfg(feature = "frequency-rococo-local")]
 pub mod frequency_local;
+
+#[cfg(feature = "frequency-rococo-testnet")]
 pub mod frequency_rococo;
```
Comment (saraswatpuneet) [inline node/service/src/chain_spec.rs:21]: selective load of specs based on feature selection

File: node/service/src/chain_spec/frequency_local.rs [hunk 6]
```diff
@@ -21,73 +21,15 @@ pub fn template_session_keys(keys: AuraId) -> frequency_rococo_runtime::SessionK
...
-		None,
-		None,
-		None,
-		Extensions {
-			relay_chain: "rococo-local".into(), // You MUST set this to the correct network!
-			para_id: 2000,
-		},
-	)
-}
-
```
Comment (saraswatpuneet) [inline node/service/src/chain_spec/frequency_local.rs:81]: We use instant sealing or rococ-local locally so we only need one of the two

File: scripts/generate_specs.sh [hunk 7]
```diff
@@ -19,7 +19,7 @@ case $build_step in
   rococo-2000)
     mkdir -p ./res/genesis/local
     echo "Building Spec for  frequency rococo localnet paraid=2000"
-    $PWD/target/release/frequency build-spec --disable-default-bootnode > ./res/genesis/local/frequency-spec-rococo.json
+    $PWD/target/release/frequency build-spec --disable-default-bootnode --chain=frequency-local > ./res/genesis/local/frequency-spec-rococo.json
     sed -i.bu "s/\"parachainId\": 2000/\"parachainId\": $parachain_id/g" ./res/genesis/local/frequency-spec-rococo.json
     sed -i.bu "s/\"para_id\": 2000/\"para_id\": $parachain_id/g" ./res/genesis/local/frequency-spec-rococo.json
     $PWD/target/release/frequency build-spec --raw --disable-default-bootnode --chain ./res/genesis/local/frequency-spec-rococo.json > ./res/genesis/local/rococo-local-frequency-2000-raw.json
```
Comment (saraswatpuneet) [inline scripts/generate_specs.sh:25]: now we want to specify which chain id we are building

Comment (saraswatpuneet) [reply scripts/generate_specs.sh:25]: For local runs now we will never need specs to be generated

File: scripts/init.sh [hunk 8]
```diff
@@ -142,11 +144,11 @@ onboard-frequency)
...
+      genesis=$(docker run -it {REPO_NAME}/frequency:${frequency_docker_image_tag} export-genesis-state --chain="frequency-local")
+      docker run -it {REPO_NAME}/frequency:${frequency_docker_image_tag} export-genesis-wasm --chain="frequency-local" > $wasm_location
     else
-      genesis=$(./target/release/frequency export-genesis-state --chain="${chain_spec}")
-      ./target/release/frequency export-genesis-wasm --chain="${chain_spec}" > $wasm_location
+      genesis=$(./target/release/frequency export-genesis-state --chain="frequency-local")
+      ./target/release/frequency export-genesis-wasm --chain="frequency-local" > $wasm_location
     fi
 
   echo "WASM path:" "${parachain}-${para_id}.wasm"
```
Comment (saraswatpuneet) [inline scripts/init.sh:154]: same here, generate specs/ run frequency locally using chain id, for local runs we wont need specs anymore

File: scripts/run_all_benchmarks.sh [hunk 9]
```diff
@@ -18,7 +17,7 @@ function run_benchmark() {
   $BENCHMARK \
   --pallet $1 \
   --extrinsic "*" \
-  --chain="$CHAIN" \
+  --chain="frequency-rococo" \
```
Comment (wilwade) [inline scripts/run_all_benchmarks.sh]: This feels wrong.

File: scripts/init.sh (hunk 8 above)
Comment (wilwade) [reply scripts/init.sh:154]: Are there any readme updates around this?

File: node/cli/Cargo.toml [hunk 10]
```diff
@@ -57,9 +55,13 @@ cli = [
   "frame-benchmarking-cli",
   "try-runtime-cli",
 ]
-default = ["wasmtime", "cli"]
+default = ["std", "wasmtime", "cli", "all-frequency-features"]
 runtime-benchmarks = [
```
Comment (saraswatpuneet) [inline node/cli/Cargo.toml:59]: if someone runs cargo check or cargo build , default all features are built

**Review by wilwade — CHANGES_REQUESTED** [review]
Two things and a blocker to check with @demisx around Docker file building

File: scripts/run_all_benchmarks.sh (hunk 9 above)
Comment (saraswatpuneet) [reply scripts/run_all_benchmarks.sh]: @wilwade  I have changed it to frequency , you may need to refesh

File: scripts/init.sh (hunk 8 above)
Comment (saraswatpuneet) [reply scripts/init.sh:154]: Not really these are internal working of frequency the functionality to run frequency locally remains the same be it sealing or rococo-local

Comment (saraswatpuneet) [reply scripts/init.sh:154]: Also after this PR it is not required to generate specs locally the chain flag will use the id to generate on the fly

File: scripts/generate_specs.sh [hunk 11]
```diff
@@ -4,13 +4,13 @@ set -e
 
 parachain_id=$1
 build_step=$2
-
+profile=$3
 if [[ $parachain_id == "" ]]; then
   echo "Chain Name or Parachain ID argument not provided"
```
Comment (saraswatpuneet) [inline scripts/generate_specs.sh:9]: @wilwade  here the profile is passed as parameters and makefile/spec job has been updated

File: scripts/run_all_benchmarks.sh [hunk 12]
```diff
@@ -3,10 +3,9 @@
 export RUST_LOG=info
 THIS_DIR=$( dirname -- "$0"; )
 PROJECT=${1:-$THIS_DIR/..}
-RUNTIME=$PROJECT/target/release/frequency
+RUNTIME=$PROJECT/target/production/frequency
 BENCHMARK="$RUNTIME benchmark pallet "
```
Comment (saraswatpuneet) [inline scripts/run_all_benchmarks.sh:7]: @wilwade  here the benchmarks are always run on production binary so this fixes the benchmarks

File: Makefile (hunk 3 above)
Comment (shannonwells) [reply Makefile:106]: I wonder if it makes sense to add Make rules for the production target spec generation

File: node/service/src/chain_spec/frequency_local.rs [hunk 13]
```diff
@@ -21,73 +21,15 @@ pub fn template_session_keys(keys: AuraId) -> frequency_rococo_runtime::SessionK
 	frequency_rococo_runtime::SessionKeys { aura: keys }
 }
 
-pub fn development_config() -> ChainSpec {
```
Comment (shannonwells) [inline node/service/src/chain_spec/frequency_local.rs:24]: Where does a development genesis config live now?

Comment (saraswatpuneet) [reply node/service/src/chain_spec/frequency_local.rs:24]: It was same as testnet, got rid of one as we either run frequency as instan seal or as rococo collator locally

File: Makefile (hunk 3 above)
Comment (saraswatpuneet) [reply Makefile:106]: it is used in a github job to generate specs and binary, mainly used for our releases unless someone wants to check things out locally

File: node/Cargo.toml [hunk 14]
```diff
@@ -27,7 +27,7 @@ substrate-build-script-utils = {git = "https://github.com/paritytech/substrate",
 
 [features]
 std = []
-default = ["std", "all-frequency-features"]
+default = ["std"]
 runtime-benchmarks = [
```
Comment (saraswatpuneet) [inline node/Cargo.toml:31]: default will always build all features hence we dont want to put it in default

**Review by wilwade — APPROVED** [review]
🥇

File: node/cli/src/command.rs [hunk 15]
```diff
@@ -100,11 +200,41 @@ impl SubstrateCli for Cli {
 	}
 
 	fn native_runtime_version(spec: &Box<dyn ChainSpec>) -> &'static RuntimeVersion {
-		match spec.identify() {
-			ChainIdentity::Frequency => &frequency_runtime::VERSION,
-			ChainIdentity::FrequencyRococo => &frequency_rococo_runtime::VERSION,
-			ChainIdentity::FrequencyLocal => &frequency_rococo_runtime::VERSION,
+		if spec.id() == "frequency" {
+			if cfg!(feature = "frequency") {
```
Comment (aramikm) [inline node/cli/src/command.rs:204]: IMO This looks too messy, there might be a way to organize these in a cleaner way. Generally speaking most of changes on `command.rs` might have some potential for improvement.

One question might worth asking is that how much we want to optimize the binary. Like there might be an argument to include some of those lines for the sake of readability if added number of lines are not that significant.

**Review by aramikm — COMMENTED** [review]
Great job, added one comment.

File: node/cli/src/command.rs (hunk 15 above)
Comment (saraswatpuneet) [reply node/cli/src/command.rs:204]: @aramikm  yes, there will be another PR for this story that will further cleanup

**Pull request merged**